athena-docdb/src/main/java/com/amazonaws/athena/connectors/docdb/DocDBRecordHandler.java

```diff
@@ -127,74 +141,147 @@ private static Map<String, Object> documentAsMap(Document document, boolean case
     }
 
     /**
-     * Scans DocumentDB using the scan settings set on the requested Split by DocDBeMetadataHandler.
+     * Scans DocumentDB using the scan settings set on the requested Split by DocDBMetadataHandler.
+     * This method handles query execution with various optimizations including predicate pushdown,
+     * limit pushdown, sort pushdown, and projection optimization.
      *
+     * @param spiller The BlockSpiller to write results to
+     * @param recordsRequest The ReadRecordsRequest containing query details and constraints
+     * @param queryStatusChecker Used to check if the query is still running
      * @see RecordHandler
      */
     @Override
     protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recordsRequest, QueryStatusChecker queryStatusChecker)
     {
-        TableName tableNameObj = recordsRequest.getTableName();
-        String schemaName = tableNameObj.getSchemaName();
-        String tableName = recordsRequest.getSchema().getCustomMetadata().getOrDefault(
+        final TableName tableNameObj = recordsRequest.getTableName();
+        final String schemaName = tableNameObj.getSchemaName();
+        final String tableName = recordsRequest.getSchema().getCustomMetadata().getOrDefault(
             SOURCE_TABLE_PROPERTY, tableNameObj.getTableName());
 
-        logger.info("Resolved tableName to: {}", tableName);
-        Map<String, ValueSet> constraintSummary = recordsRequest.getConstraints().getSummary();
+        logger.info("Starting readWithConstraint for schema: {}, table: {}", schemaName, tableName);
+        
+        final Map<String, ValueSet> constraintSummary = recordsRequest.getConstraints().getSummary();
+        logger.info("Processing {} constraints", constraintSummary.size());
 
-        MongoClient client = getOrCreateConn(recordsRequest.getSplit());
-        MongoDatabase db;
-        MongoCollection<Document> table;
+        final MongoClient client = getOrCreateConn(recordsRequest.getSplit());
+        final MongoDatabase db;
+        final MongoCollection<Document> table;
         Document query;
 
+        // ---------------------- Substrait Plan extraction ----------------------
```

Comment:
can we move this logic to a util function or a new class that returns the desired predicates with the constraints object as the input?